File: backend/app/schemas/practice.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


PracticeStatus = Literal["not_started", "practicing", "ready_for_retest", "completed"]


class PracticeFromReportCreate(BaseModel):
    report_id: int = Field(gt=0)
    weakness_key: str = Field(min_length=1, max_length=255)
    weakness_title: str = Field(min_length=1, max_length=500)


class PracticeFromQuestionReviewCreate(PracticeFromReportCreate):
    question_review_id: int = Field(gt=0)
    score_id: int = Field(gt=0)
    question_index: int = Field(gt=0)
    practice_mode: Literal["repeat_question", "similar_question"]


class PracticeCreationRead(BaseModel):
    id: int
    status: PracticeStatus
    practice_session_id: int | None = None
    retest_session_id: int | None = None

    model_config = {"from_attributes": True}


class PracticeListItem(PracticeCreationRead):
    source_report_id: int
    source_session_id: int
    source_question_review_id: int | None = None
    weakness_key: str
    weakness_title: str
    target_dimension: str | None = None
    before_score: int | None = None
    after_score: int | None = None
    created_at: datetime
    updated_at: datetime


class PracticeComparisonSession(BaseModel):
    session_id: int
    score: int
    weaknesses: list[str] = Field(default_factory=list)
    answer: str = ""
    answer_structure: list[str] = Field(default_factory=list)
    sub_scores: dict[str, int] = Field(default_factory=dict)


class PracticeComparisonDelta(BaseModel):
    score: int = 0
    resolved_weaknesses: list[str] = Field(default_factory=list)
    remaining_weaknesses: list[str] = Field(default_factory=list)
    sub_scores: dict[str, int] = Field(default_factory=dict)


class PracticeNextWeakness(BaseModel):
    key: str
    title: str


class PracticeComparisonRead(BaseModel):
    practice_id: int
    status: PracticeStatus
    source_report_id: int
    practice_session_id: int | None = None
    retest_session_id: int | None = None
    weakness_title: str
    before: PracticeComparisonSession
    after: PracticeComparisonSession | None = None
    delta: PracticeComparisonDelta
    summary: str
    next_weakness: PracticeNextWeakness | None = None

File: frontend/src/api/practice.ts
import { apiClient } from './client'
import { normalizeInterviewSession, type InterviewSession } from './interview'

export type QuestionPracticeMode = 'repeat_question' | 'similar_question'
export type PracticeStatus = 'not_started' | 'practicing' | 'ready_for_retest' | 'completed'

export interface PracticeCreationResponse {
  id: number
  status: PracticeStatus
  practice_session_id?: number | null
  retest_session_id?: number | null
}

export interface QuestionPracticeInput {
  report_id: number
  question_review_id: number
  score_id: number
  question_index: number
  weakness_key: string
  weakness_title: string
  practice_mode: QuestionPracticeMode
}

export interface ReportPracticeInput {
  report_id: number
  weakness_key: string
  weakness_title: string
}

export interface PracticeListItem extends PracticeCreationResponse {
  source_report_id: number
  source_session_id: number
  source_question_review_id?: number | null
  weakness_key: string
  weakness_title: string
  target_dimension?: string | null
  before_score?: number | null
  after_score?: number | null
  created_at?: string
  updated_at?: string
}

export interface PracticeComparisonSession {
  session_id: number
  score: number
  weaknesses: string[]
  answer: string
  answer_structure: string[]
  sub_scores: Record<string, number>
}

export interface PracticeComparisonDelta {
  score: number
  resolved_weaknesses: string[]
  remaining_weaknesses: string[]
  sub_scores: Record<string, number>
}

export interface PracticeNextWeakness {
  key: string
  title: string
}

export interface PracticeComparison {
  practice_id: number
  status: PracticeStatus
  source_report_id?: number | null
  practice_session_id?: number | null
  retest_session_id?: number | null
  weakness_title?: string | null
  before: PracticeComparisonSession
  after: PracticeComparisonSession | null
  delta: PracticeComparisonDelta
  summary: string
  next_weakness?: PracticeNextWeakness | null
}

export function createPracticeFromQuestionReview(data: QuestionPracticeInput) {
  return apiClient.post<PracticeCreationResponse>('/practice/from-question-review', data)
}

export function createPracticeFromReport(data: ReportPracticeInput) {
  return apiClient.post<PracticeCreationResponse>('/practice/from-report', data)
}

export function fetchPractices() {
  return apiClient.get<PracticeListItem[]>('/practice')
}

export async function startPractice(id: number) {
  const response = await apiClient.post<InterviewSession>(`/practice/${id}/start`)
  return { ...response, data: normalizeInterviewSession(response.data) }
}

export async function startPracticeRetest(id: number) {
  const response = await apiClient.post<InterviewSession>(`/practice/${id}/start-retest`)
  return { ...response, data: normalizeInterviewSession(response.data) }
}

export async function fetchPracticeComparison(id: number) {
  const response = await apiClient.get<PracticeComparison>(`/practice/${id}/comparison`)
  return { ...response, data: normalizePracticeComparison(response.data, id) }
}

function normalizePracticeComparison(data: PracticeComparison, practiceId: number): PracticeComparison {
  const before = normalizeComparisonSession(data.before)
  const after = data.after ? normalizeComparisonSession(data.after) : null
  const fallbackDelta = after ? after.score - before.score : 0
  return {
    ...data,
    practice_id: Number(data.practice_id || practiceId),
    status: after ? 'completed' : data.status || 'ready_for_retest',
    before,
    after,
    delta: {
      score: finiteNumber(data.delta?.score, fallbackDelta),
      resolved_weaknesses: stringList(data.delta?.resolved_weaknesses),
      remaining_weaknesses: stringList(data.delta?.remaining_weaknesses),
      sub_scores: numberRecord(data.delta?.sub_scores)
    },
    summary: data.summary || ''
  }
}

function normalizeComparisonSession(data: PracticeComparisonSession): PracticeComparisonSession {
  return {
    ...data,
    session_id: Number(data.session_id),
    score: finiteNumber(data.score),
    weaknesses: stringList(data.weaknesses),
    answer: data.answer || '',
    answer_structure: stringList(data.answer_structure),
    sub_scores: numberRecord(data.sub_scores)
  }
}

function stringList(value: unknown) {
  return Array.isArray(value)
    ? value.filter((item): item is string => typeof item === 'string' && Boolean(item.trim()))
    : []
}

function numberRecord(value: unknown) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return {}
  return Object.fromEntries(
    Object.entries(value)
      .map(([key, score]) => [key, Number(score)] as const)
      .filter(([, score]) => Number.isFinite(score))
  )
}

function finiteNumber(value: unknown, fallback = 0) {
  const number = Number(value)
  return Number.isFinite(number) ? number : fallback
}
